File: core-register/src/test/java/com/jeanbarrossilva/memento/core/register/editor/EditorTests.java
package com.jeanbarrossilva.memento.core.register.editor;

import static org.assertj.core.api.Assertions.assertThat;

import com.jeanbarrossilva.memento.core.register.TestRegister;
import com.jeanbarrossilva.memento.core.register.domain.Color;
import com.jeanbarrossilva.memento.core.register.infra.Editor;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.util.concurrent.ExecutionException;

public class EditorTests {
    @NotNull
    private final TestRegister register = new TestRegister();

    @NotNull
    private final Editor editor = new TestEditor(register);

    @Nullable
    private String noteID;

    @BeforeEach
    void setUp() throws ExecutionException, InterruptedException {
        register
            .register("Title", "Body", Color.BLUE)
            .thenAccept((noteID) -> this.noteID = noteID)
            .get();
    }

    @AfterEach
    void tearDown() throws ExecutionException, InterruptedException {
        assert noteID != null;
        register.unregister(noteID).get();
    }

    @Test
    void setTitle() throws ExecutionException, InterruptedException {
        assert noteID != null;
        editor.setTitle(noteID, "New title").thenCompose(result -> register
            .getNoteByID(noteID)
            .thenAcceptAsync(note -> assertThat(note.getTitle()).isEqualTo("New title")))
            .get();
    }

    @Test
    void setBody() throws ExecutionException, InterruptedException {
        assert noteID != null;
        editor
            .setBody(noteID, "New body")
            .thenCompose(result -> register
                .getNoteByID(noteID)
                .thenAcceptAsync(note -> assertThat(note.getBody()).isEqualTo("New body")))
            .get();
    }
}

File: core-register/src/test/java/com/jeanbarrossilva/memento/core/register/editor/EditorTests.kt
package com.jeanbarrossilva.memento.core.register.editor

import app.cash.turbine.test
import com.jeanbarrossilva.memento.core.register.TestRegister
import com.jeanbarrossilva.memento.core.register.domain.Color
import com.jeanbarrossilva.memento.core.register.repository.TestRepository
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.runTest
import kotlin.test.AfterTest
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals

internal class EditorTests {
    private val repository = TestRepository()
    private val register = TestRegister(repository)
    private val editor = TestEditor(repository)
    private lateinit var noteID: String

    @BeforeTest
    @OptIn(ExperimentalCoroutinesApi::class)
    fun setUp() {
        runTest {
            noteID = register.register("Title", body = "Body")
        }
    }

    @AfterTest
    @OptIn(ExperimentalCoroutinesApi::class)
    fun tearDown() {
        runTest {
            register
        }
    }

    @Test
    @OptIn(ExperimentalCoroutinesApi::class)
    fun `GIVEN a note WHEN setting its title THEN it's set`() {
        runTest {
            editor.setTitle(noteID, "New title")
            repository.getNoteByID(noteID).test { assertEquals("New title", awaitItem()?.title) }
        }
    }

    @Test
    @OptIn(ExperimentalCoroutinesApi::class)
    fun `GIVEN a note WHEN setting its body THEN it's set`() {
        runTest {
            editor.setBody(noteID, "New body")
            repository.getNoteByID(noteID).test { assertEquals("New body", awaitItem()?.body) }
        }
    }

    @Test
    @OptIn(ExperimentalCoroutinesApi::class)
    fun `GIVEN a note WHEN setting its color THEN it's set`() {
        runTest {
            editor.setColor(noteID, Color.YELLOW)
            repository.getNoteByID(noteID).test { assertEquals(Color.YELLOW, awaitItem()?.color) }
        }
    }
}
